src/balancer-strategy-manager/BalancerV2WeightedPoolStrategyManager.sol

```diff
@@ -0,0 +1,226 @@
+import {Rescuable} from 'solidity-utils/contracts/utils/Rescuable.sol';
+import {RescuableBase, IRescuableBase} from 'solidity-utils/contracts/utils/RescuableBase.sol';
+
+import {IBalancerPool} from './balancer-v2/IBalancerPool.sol';
+import {IBalancerVault, IAsset} from './balancer-v2/IBalancerVault.sol';
+import {WeightedPoolUserData} from './balancer-v2/WeightedPoolUserData.sol';
+import {IBalancerStrategyManager} from './IBalancerStrategyManager.sol';
+
+contract BalancerV2WeightedPoolStrategyManager is
+  IBalancerStrategyManager,
+  OwnableWithGuardian,
+  Rescuable
+{
+  using SafeERC20 for IERC20;
+
+  bytes32 public immutable POOL_ID;
+  IBalancerPool public immutable POOL;
+  IBalancerVault public immutable VAULT;
+  uint256 public immutable TOKEN_COUNT;
+  address public immutable HYPERNATIVE;
+
+  mapping(uint256 => TokenConfig) private tokenConfig;
+
+  modifier onlyWithdrawable() {
+    if (_msgSender() != owner() && _msgSender() != guardian() && _msgSender() != HYPERNATIVE) {
+      revert AccessForbidden();
+    }
+    _;
+  }
+
+  constructor(
+    address _vault,
+    bytes32 _poolId,
+    TokenConfig[] memory _tokenConfig,
+    address _owner,
+    address _guardian,
+    address _hypernative
+  ) {
+    VAULT = IBalancerVault(_vault);
+    POOL_ID = _poolId;
+
+    (address poolAddress, ) = VAULT.getPool(_poolId);
+    POOL = IBalancerPool(poolAddress);
+
+    address[] memory actuallTokens;
+    (actuallTokens, , ) = VAULT.getPoolTokens(_poolId);
+
+    TOKEN_COUNT = actuallTokens.length;
+
+    if (_tokenConfig.length != TOKEN_COUNT) {
+      revert TokenCountMismatch();
+    }
+
+    for (uint256 i = 0; i < TOKEN_COUNT; ) {
+      if (actuallTokens[i] != _tokenConfig[i].token) {
+        revert TokenMismatch();
+      }
+
+      tokenConfig[i] = _tokenConfig[i];
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    _transferOwnership(_owner);
+    _updateGuardian(_guardian);
+    HYPERNATIVE = _hypernative;
+  }
+
+  /**
+   * @dev sets token provider
+   * @param _id The index of token to set
+   * @param _provider The address of provider
+   */
+  function setTokenProvider(uint256 _id, address _provider) external onlyOwner {
+    if (_id >= TOKEN_COUNT) {
+      revert TokenCountMismatch();
+    }
+
+    tokenConfig[_id].provider = _provider;
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function deposit(uint256[] calldata _tokenAmounts) external onlyOwnerOrGuardian {
+    if (_tokenAmounts.length != TOKEN_COUNT) {
+      revert TokenCountMismatch();
+    }
+
+    for (uint256 i = 0; i < TOKEN_COUNT; ) {
+      IERC20 token = IERC20(tokenConfig[i].token);
+
+      if (_tokenAmounts[i] > token.balanceOf(address(this))) {
+        revert InsufficientToken(address(token));
+      }
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    bytes memory userData = abi.encode(WeightedPoolUserData.JoinKind.EXACT_TOKENS_IN_FOR_BPT_OUT);
```

Comment:
You are not using this variable for anything until below, you can just set there